File: db/evolution.py
"""The Generation Evolution Explorer's one Module: a READ-ONLY view of
a StoryPlan.

`load(conn, plan_id, ...)` assembles everything the page shows from
things that already exist and decides nothing:

    StoryPlan        phases, families, chronology supported or not,
                     representatives, claims              -- AUTHORITY
    StorySnapshot    members, prompt texts by role, parameters,
                     artifacts, lineage, dates             -- AUTHORITY
    prompt vectors   db/prompts.py, by FROZEN main-section text hash
                     under one (space, query policy)
    media vectors    derived_embedding, by the member's frozen uuid AND
                     frozen content_sha256 -- a replaced file's vector
                     is never substituted
    lineage          the snapshot's frozen parent/child edges

Over those frozen facts it MEASURES: cosine between consecutive members'
effective prompts and images (sequenced plans only -- O(n) pairs, never
a matrix), cosine between each member's prompt as written and as run,
cosine between each effective prompt and its image, and exact parameter
deltas. Every metric names one space and one policy; nothing blends
spaces into a score; every unavailable metric says why. Phase
boundaries are drawn where the plan put them, never where a cosine
dipped. No writes, no model loads, no embedding computation: the same
plan over the same cached vectors yields byte-identical JSON.
"""

from __future__ import annotations

import datetime
import itertools
import typing

from . import planning, prompts, stories

FORMAT_VERSION = 1

#: The recipe facts a transition compares, one per column of db/schema.sql
#: `generation`. Closed: a fact this list does not name is not compared,
#: and one it names that the table does not hold reads as None everywhere.
Parameter = typing.Literal["seed", "steps", "cfg", "denoise", "clip_skip", "sampler", "scheduler", "width", "height"]

#: What a transition can report as changed: a recipe parameter, or the
#: checkpoint. LoRAs are not here -- they are identities that come and go,
#: reported as added and removed rather than as before and after.
ChangedParameter = typing.Literal[Parameter, "model"]

_PARAMS: tuple[Parameter, ...] = typing.get_args(Parameter)


def _cosine(a, b) -> float:
    from .similarity import normalise

    unit = normalise([a, b])
    return round(float(unit[0] @ unit[1]), 4)


def _space(conn, provider: str | None, models_dir: str) -> dict:
    """The one (space, policy) every metric is measured in: the named
    provider's joint space and current query policy, or the first
    configured provider's. Nothing loads; an unminted space is reported,
    not created."""
    from vision import semantic

    from . import retrieval

    held: dict[str, str | int | None] = {
        "provider": None,
        "space_id": None,
        "space": None,
        "prompt_policy_hash": None,
        "unavailable": None,
    }
    if provider is not None:
        name, model, configured = retrieval.choice_for(conn, provider)  # exact; ambiguity refused
    else:
        try:
            choices = retrieval.choices(conn)
        except ValueError as refused:
            held["unavailable"] = str(refused)
            return held
        if len(choices) != 1:
            held["unavailable"] = "several semantic spaces are configured; name one with ?space=" + ", ".join(
                retrieval.spelled(*one) for one in choices
            )
            return held
        name, model, configured = choices[0]
    checkpoint = semantic.pin(name, models_dir, model, configured)
    held["provider"] = name
    held["space"] = semantic.space(name, model, checkpoint, 1).key
    held["prompt_policy_hash"] = semantic.policy_hash(name, model, checkpoint)
    found = prompts.space_of(conn, name, model, checkpoint)
    if found is None:
        held["unavailable"] = f"no vectors recorded under {held['space']} yet; run /jobs/embed and /jobs/embed_prompts"
    else:
        held["space_id"] = found[0]
    return held


def _media_vectors(conn, sid: int | None, members: list[dict]) -> dict[str, list[float]]:
    """Media vectors by member ref -- only a row computed from the
    FROZEN bytes (source_sha256 = the snapshot's content_sha256) for
    the frozen file identity counts."""
    import numpy as np

    if sid is None or not members:
        return {}
    held: dict[str, list[float]] = {}
    by_uuid = {one["file_uuid"]: one for one in members}
    marks = ",".join("?" for _ in by_uuid)
    for uuid, sha, blob in conn.execute(
        "SELECT ent.uuid, e.source_sha256, e.vector FROM derived_embedding e"
        " JOIN entity ent ON ent.id = e.file_id"
        f" WHERE e.space_id = ? AND ent.uuid IN ({marks})",
        (sid, *[bytes.fromhex(u) for u in by_uuid]),
    ):
        member = by_uuid.get(uuid.hex())
        if member is not None and sha == member["content_sha256"]:
            held[planning._member_ref(member["ordinal"])] = [float(x) for x in np.frombuffer(blob, dtype=np.float32)]
    return held


def _slugs(conn, members: list[dict]) -> dict[str, str | None]:
    """Each member's CURRENT address by its frozen uuid -- presentation
    only; None when the file is gone. Batched: the statement count is
    bounded however large the session."""
    held: dict[str, str | None] = {planning._member_ref(one["ordinal"]): None for one in members}
    by_uuid = {one["file_uuid"]: planning._member_ref(one["ordinal"]) for one in members}
    uuids = sorted(by_uuid)
    for start in range(0, len(uuids), 500):
        piece = uuids[start : start + 500]
        marks = ",".join("?" for _ in piece)
        for uuid, kind, slug in conn.execute(
            f"SELECT uuid, kind, slug FROM entity WHERE uuid IN ({marks})", [bytes.fromhex(u) for u in piece]
        ):
            if kind == "file":
                held[by_uuid[uuid.hex()]] = slug
    return held


def _prompt_ids(conn, hashes: set[str]) -> dict[str, int]:
    """Prompt rows by frozen text hash -- the one live lookup the page
    needs for its "prompts like this" link; absent when no row holds
    the text any more."""
    if not hashes:
        return {}
    wanted = sorted(hashes)
    marks = ",".join("?" for _ in wanted)
    return {
        row[0]: int(row[1])
        for row in conn.execute(f"SELECT text_hash, id FROM prompt WHERE text_hash IN ({marks})", wanted)
    }


def _prompt_view(frozen: dict | None, prompt_ids: dict[str, int]) -> dict | None:
    """One role's frozen prompt as the page reads it, or None when the
    snapshot froze none."""
    if frozen is None:
        return None
    return {
        "text": frozen["text"],
        "hash": frozen["hash"],
        "main": frozen["main"],
        "main_hash": frozen["main_hash"],
        "prompt_id": prompt_ids.get(frozen["main_hash"]),
    }


def _generation_facts(generation: dict | None) -> dict:
    generation = generation or {}
    artifacts = generation.get("artifacts") or []
    return {
        **{key: generation.get(key) for key in _PARAMS},
        "tool": generation.get("tool"),
        "model": next((a["name"] for a in artifacts if a["role"] == "checkpoint"), None),
        "loras": sorted(a["name"] for a in artifacts if a["role"] == "lora"),
        "lora_uuids": sorted(a["uuid"] for a in artifacts if a["role"] == "lora"),
        "lora_names": {a["uuid"]: a["name"] for a in artifacts if a["role"] == "lora"},
    }


def _changes(before: dict, after: dict) -> dict:
    """Exact deltas. LoRAs differ by frozen IDENTITY (uuid), spelled by
    their frozen names: two different files that share a name changed,
    one file renamed did not."""
    changed = {key: {"from": before[key], "to": after[key]} for key in (*_PARAMS, "model") if before[key] != after[key]}
    named = {**before["lora_names"], **after["lora_names"]}
    added = sorted(set(after["lora_uuids"]) - set(before["lora_uuids"]))
    removed = sorted(set(before["lora_uuids"]) - set(after["lora_uuids"]))
    return {
        **changed,
        "loras_added": [named[uuid] for uuid in added],
        "loras_removed": [named[uuid] for uuid in removed],
        "lora_uuids_added": added,
        "lora_uuids_removed": removed,
    }


def local_day(snapshot: dict) -> str | None:
    """The session's LOCAL calendar day as the gallery's day facet spells
    it, when the event has a wall clock -- an identity; the web adapter
    turns it into an address."""
    when = snapshot["subject"]["time"].get("local")
    if not when:
        return None
    return datetime.datetime.fromtimestamp(when[0], datetime.UTC).strftime("%Y-%m-%d")


def load(conn, plan_id: int, *, provider: str | None = None, models_dir: str = "") -> dict:
    """The EvolutionView for one plan. Raises LookupError for an unknown
    plan and stories.Corrupt for one that no longer verifies."""
    row = conn.execute("SELECT snapshot_id, document_sha256 FROM story_plan WHERE id = ?", (plan_id,)).fetchone()
    if row is None:
        raise LookupError(f"no story plan {plan_id}")
    plan = planning.load_plan(conn, plan_id)
    snapshot = stories.load_snapshot(conn, int(row[0]))
    members = sorted(snapshot["members"], key=lambda one: one["ordinal"])
    refs = [planning._member_ref(one["ordinal"]) for one in members]
    sequenced = bool(plan["subject"]["sequenced"])
    phase_of = {ref: phase["id"] for phase in plan["phases"] for ref in phase["member_refs"]}
    semantic = _space(conn, provider, models_dir)
    sid, policy = semantic["space_id"], semantic["prompt_policy_hash"]

    # prompts as the planner reads them: the MAIN section under the tool's grammar
    # prompts as the snapshot FROZE them: the main section read when the
    # evidence was frozen, never re-read by today's parser
    texts: dict[str, dict[str, dict | None]] = {}
    for ref, one in zip(refs, members, strict=True):
        generation = one.get("generation") or {}
        by_role = {p["role"]: p for p in generation.get("prompts") or []}
        held: dict[str, dict | None] = {}
        for role in ("effective", "original"):
            frozen = by_role.get(role)
            if frozen is None:
                held[role] = None
                continue
            main = planning.frozen_main(one, role)
            held[role] = {
                "text": frozen["text"],
                "hash": frozen["text_hash"],
                "main": main,
                "main_hash": frozen.get("main_hash") or prompts.text_hash(main),
            }
        texts[ref] = held
    hashes = {one["main_hash"] for held in texts.values() for one in held.values() if one}
    prompt_vectors = prompts.current_vectors(conn, sid, policy, sorted(hashes)) if sid is not None else {}
    media_vectors = _media_vectors(conn, sid, members)
    slugs = _slugs(conn, members)
    prompt_ids = _prompt_ids(conn, hashes)
    no_space = semantic["unavailable"]

    def prompt_vector(ref: str, role: str):
        held = texts[ref][role]
        if held is None:
            return None, f"no frozen {role} prompt"
        if no_space:
            return None, no_space
        vector = prompt_vectors.get(held["main_hash"])
        if vector is None:
            return (
                None,
                "no current vector for the frozen prompt text under this space and policy; run /jobs/embed_prompts",
            )
        return vector, None

    def media_vector(ref: str):
        if no_space:
            return None, no_space
        vector = media_vectors.get(ref)
        if vector is None:
            return None, "no media vector computed from the frozen bytes under this space; run /jobs/embed"
        return vector, None

    def metric(name: str, a, why_a, b, why_b) -> dict:
        if a is None or b is None:
            return {name: None, f"{name}_unavailable": why_a or why_b}
        return {name: _cosine(a, b)}

    facts = {ref: _generation_facts(one.get("generation")) for ref, one in zip(refs, members, strict=True)}
    out_members = []
    for ref, one in zip(refs, members, strict=True):
        effective, why_e = prompt_vector(ref, "effective")
        original, why_o = prompt_vector(ref, "original")
        image, why_i = media_vector(ref)
        held = texts[ref]
        slug = slugs[ref]
        out_members.append(
            {
                "ref": ref,
                "phase_ref": phase_of.get(ref),
                "media": {
                    "uuid": one["file_uuid"],
                    "name": one["name"],
                    "kind": one["media_kind"],
                    "content_sha256": one["content_sha256"],
                    "slug": slug,
                },
                "occurrence": one.get("occurrence"),
                "prompt": {role: _prompt_view(held[role], prompt_ids) for role in ("effective", "original")},
                "generation": {key: value for key, value in facts[ref].items() if key != "lora_names"},
                "metrics": {
                    **metric("original_effective_cosine", original, why_o, effective, why_e),
                    **metric("text_image_cosine", effective, why_e, image, why_i),
                },
            }
        )

    transitions = []
    if sequenced:
        for before, after in itertools.pairwise(refs):
            a_prompt, why_ap = prompt_vector(before, "effective")
            b_prompt, why_bp = prompt_vector(after, "effective")
            a_image, why_ai = media_vector(before)
            b_image, why_bi = media_vector(after)
            transitions.append(
                {
                    "from": before,
                    "to": after,
                    "phase_boundary": phase_of.get(before) != phase_of.get(after),
                    **metric("prompt_cosine", a_prompt, why_ap, b_prompt, why_bp),
                    **metric("visual_cosine", a_image, why_ai, b_image, why_bi),
                    "changes": _changes(facts[before], facts[after]),
                }
            )

    by_uuid = {one["file_uuid"]: ref for ref, one in zip(refs, members, strict=True)}
    lineage = []
    seen = set()
    for ref, one in zip(refs, members, strict=True):
        for edge in (one.get("lineage") or {}).get("parents") or []:
            key = (edge["uuid"], ref, edge["kind"])
            if key not in seen:
                seen.add(key)
                lineage.append({"parent": by_uuid.get(edge["uuid"], edge["uuid"]), "child": ref, "kind": edge["kind"]})
        for edge in (one.get("lineage") or {}).get("children") or []:
            key = (ref, edge["uuid"], edge["kind"])
            if key not in seen and edge["uuid"] not in by_uuid:
                seen.add(key)
                lineage.append({"parent": ref, "child": edge["uuid"], "kind": edge["kind"]})

    claims = {claim["id"]: claim for claim in plan["claims"]}
    phases = [
        {
            "id": phase["id"],
            "label": phase["label_hint"],
            "member_refs": list(phase["member_refs"]),
            "representative_refs": list(phase["representative_refs"]),
            "claims": [
                {"id": cid, "kind": claims[cid]["kind"], "facts": claims[cid]["facts"]} for cid in phase["claim_refs"]
            ],
        }
        for phase in plan["phases"]
    ]
    return {
        "v": FORMAT_VERSION,
        "plan": {
            "id": plan_id,
            "sha256": row[1],
            "format": plan["v"],
            "sequenced": sequenced,
            "unsupported": plan["unsupported"],
            "label": plan["subject"]["label_hint"],
        },
        "snapshot": {
            "sha256": plan["snapshot_sha256"],
            "time": snapshot["subject"]["time"],
            "members": len(members),
            "subject": snapshot["subject"]["event_kind"],
        },
        "semantic": semantic,
        "phases": phases,
        "members": out_members,
        "transitions": transitions,
        "lineage": lineage,
        "identities": {"local_day": local_day(snapshot)},
    }

File: sg_web/static/evolution.js
// The Generation Evolution Explorer page: several presentations of ONE
// EvolutionView (embedded JSON from db/evolution.py). This file draws;
// it decides nothing -- phases, families and chronology come from the
// plan, every number from the view. No fetches, no writes.
(() => {
  const root = document.querySelector("[data-evolution]");
  if (!root) return;
  const view = JSON.parse(root.querySelector("[data-evolution-view]").textContent);
  const main = root.querySelector("[data-main]");
  const selectedPane = root.querySelector("[data-selected]");
  const inspector = root.querySelector("[data-inspector]");
  const members = new Map(view.members.map((m) => [m.ref, m]));
  const transitionTo = new Map(view.transitions.map((t) => [t.to, t]));
  const state = { tab: view.plan.sequenced ? "sequence" : "families", selected: null, pair: null };

  const esc = (s) => String(s ?? "").replace(/[&<>"']/g, (c) => ({ "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;", "'": "&#39;" })[c]);
  const pct = (v) => (v === null || v === undefined ? "—" : `${Math.round(v * 100)}%`);
  const thumb = (m, cls = "member") =>
    m.media.thumbnail
      ? `<img class="${cls}" data-ref="${m.ref}" src="${esc(m.media.thumbnail)}" alt="${esc(m.media.name)}" title="${esc(m.ref)} · ${esc(m.media.name)}">`
      : `<span class="${cls}" data-ref="${m.ref}" title="${esc(m.media.name)} (file gone)"></span>`;

  // --- token diff: longest common subsequence over whitespace tokens ----
  function diffTokens(a, b) {
    const x = a ? a.split(/\s+/) : [], y = b ? b.split(/\s+/) : [];
    const n = x.length, m = y.length;
    const L = Array.from({ length: n + 1 }, () => new Array(m + 1).fill(0));
    for (let i = n - 1; i >= 0; i--) for (let j = m - 1; j >= 0; j--) L[i][j] = x[i] === y[j] ? L[i + 1][j + 1] + 1 : Math.max(L[i + 1][j], L[i][j + 1]);
    const out = [];
    let i = 0, j = 0;
    while (i < n && j < m) {
      if (x[i] === y[j]) { out.push(esc(x[i])); i++; j++; }
      else if (L[i + 1][j] >= L[i][j + 1]) { out.push(`<del>${esc(x[i])}</del>`); i++; }
      else { out.push(`<ins>${esc(y[j])}</ins>`); j++; }
    }
    while (i < n) out.push(`<del>${esc(x[i++])}</del>`);
    while (j < m) out.push(`<ins>${esc(y[j++])}</ins>`);
    return out.join(" ");
  }

  // --- presentations -----------------------------------------------------
  function sequence() {
    const strip = view.phases
      .map((p) => `<div class="phase" data-phase="${p.id}"><h3>${esc(p.label)}</h3><div class="members">${p.member_refs.map((r) => thumb(members.get(r))).join("")}</div></div>`)
      .join("");
    const rows = [
      ["prompt vs previous", (t) => t.prompt_cosine, (t) => t.prompt_cosine_unavailable],
      ["image vs previous", (t) => t.visual_cosine, (t) => t.visual_cosine_unavailable],
    ];
    const head = `<tr><th></th>${view.members.map((m) => `<th>${esc(m.ref.replace("member-", ""))}</th>`).join("")}</tr>`;
    const body = rows
      .map(([label, get, why]) => {
        const cells = view.members.map((m, i) => {
          if (i === 0) return "<td>·</td>";
          const t = transitionTo.get(m.ref);
          const v = get(t);
          const cls = (t.phase_boundary ? "boundary " : "") + (v === null ? "unavailable" : "");
          return `<td class="${cls}" title="${esc(why(t) || "")}">${pct(v)}</td>`;
        });
        return `<tr><th>${label}</th>${cells.join("")}</tr>`;
      })
      .join("");
    const facts = ["model", "loras", "sampler", "seed"]
      .map((key) => `<tr><th>${key}</th>${view.members.map((m) => { const v = m.generation[key]; const t = transitionTo.get(m.ref); const cls = t && t.phase_boundary ? "boundary" : ""; return `<td class="${cls}">${esc(Array.isArray(v) ? v.join(", ") : v ?? "—")}</td>`; }).join("")}</tr>`)
      .join("");
    main.innerHTML = `<div class="filmstrip">${strip}</div><div class="tracks"><table>${head}${body}${facts}</table></div>`;
  }

  function drift() {
    const W = 420, H = 320, pad = 36;
    const dots = view.transitions
      .filter((t) => t.prompt_cosine !== null && t.visual_cosine !== null)
      .map((t) => {
        const x = pad + (1 - Math.max(0, t.prompt_cosine)) * (W - 2 * pad);
        const y = H - pad - (1 - Math.max(0, t.visual_cosine)) * (H - 2 * pad);
        return `<circle data-pair="${t.from}|${t.to}" cx="${x.toFixed(1)}" cy="${y.toFixed(1)}" r="6" fill="${t.phase_boundary ? "#fc6" : "#6cf"}"><title>${t.from} → ${t.to}: prompt ${pct(t.prompt_cosine)}, image ${pct(t.visual_cosine)}</title></circle>`;
      })
      .join("");
    const missing = view.transitions.length - view.transitions.filter((t) => t.prompt_cosine !== null && t.visual_cosine !== null).length;
    main.innerHTML = `<div class="drift"><svg width="${W}" height="${H}" viewBox="0 0 ${W} ${H}">
      <line x1="${pad}" y1="${H - pad}" x2="${W - pad}" y2="${H - pad}" stroke="#555"/><line x1="${pad}" y1="${pad}" x2="${pad}" y2="${H - pad}" stroke="#555"/>
      <text x="${W / 2}" y="${H - 8}" fill="#aaa" font-size="11" text-anchor="middle">prompt change from previous →</text>
      <text x="12" y="${H / 2}" fill="#aaa" font-size="11" text-anchor="middle" transform="rotate(-90 12 ${H / 2})">image change from previous →</text>
      ${dots}</svg>
      <p class="evolution-provenance">each dot is one transition; yellow dots cross a plan phase boundary${missing ? `; ${missing} transition(s) lack a vector and are not drawn` : ""}</p></div>`;
  }

  function families() {
    main.innerHTML = `<div class="families">${view.phases
      .map((p) => `<div class="family" data-phase="${p.id}"><h3>${esc(p.label)} <small>· ${p.member_refs.length}</small></h3>
        <p class="claims">${p.claims.map((c) => esc(c.kind)).join(" · ") || "no claims"}</p>
        <div class="members">${p.member_refs.map((r) => thumb(members.get(r), p.representative_refs.includes(r) ? "member hero" : "member")).join("")}</div></div>`)
      .join("")}</div>`;
  }

  function lineage() {
    if (!view.lineage.length) { main.innerHTML = `<p class="empty">no derivation edges among these images</p>`; return; }
    const children = new Map();
    view.lineage.forEach((e) => { if (!children.has(e.parent)) children.set(e.parent, []); children.get(e.parent).push(e); });
    const isChild = new Set(view.lineage.map((e) => e.child));
    const roots = [...new Set(view.lineage.map((e) => e.parent))].filter((p) => !isChild.has(p));
    const node = (ref, kind) => {
      const m = members.get(ref);
      const label = m ? thumb(m) + ` ${esc(ref)}` : `<span class="kind">outside the session</span> ${esc(ref.slice(0, 8))}`;
      const kids = (children.get(ref) || []).map((e) => node(e.child, e.kind)).join("");
      return `<li>${label}${kind ? ` <span class="kind">${esc(kind)}</span>` : ""}${kids ? `<ul>${kids}</ul>` : ""}</li>`;
    };
    main.innerHTML = `<div class="lineage"><ul>${roots.map((r) => node(r, null)).join("")}</ul></div>`;
  }

  function compare() {
    const [a, b] = state.pair || [];
    if (!a || !b) { main.innerHTML = `<p class="empty">select two images (click one, then shift-click another)</p>`; return; }
    const A = members.get(a), B = members.get(b);
    const t = view.transitions.find((x) => (x.from === a && x.to === b) || (x.from === b && x.to === a));
    const edge = view.lineage.find((e) => (e.parent === a && e.child === b) || (e.parent === b && e.child === a));
    const rows = ["model", "loras", "sampler", "steps", "cfg", "seed", "scheduler", "width", "height"]
      .map((k) => { const va = A.generation[k], vb = B.generation[k]; const same = JSON.stringify(va) === JSON.stringify(vb); return `<tr><th>${k}</th><td>${same ? "same" : `${esc(Array.isArray(va) ? va.join(", ") : va ?? "—")} → ${esc(Array.isArray(vb) ? vb.join(", ") : vb ?? "—")}`}</td></tr>`; })
      .join("");
    main.innerHTML = `<div class="compare"><div>${thumb(A, "big")}<p>${esc(A.ref)} · ${esc(A.media.name)}</p></div><div>${thumb(B, "big")}<p>${esc(B.ref)} · ${esc(B.media.name)}</p></div>
      <div class="metrics" style="grid-column: 1 / -1"><dl>
        <dt>prompt cosine (consecutive only)</dt><dd>${t ? pct(t.prompt_cosine) : "not consecutive"}</dd>
        <dt>visual cosine (consecutive only)</dt><dd>${t ? pct(t.visual_cosine) : "not consecutive"}</dd>
        <dt>lineage</dt><dd>${edge ? `${esc(edge.parent)} → ${esc(edge.child)} (${esc(edge.kind)})` : "no derivation edge"}</dd>
        <dt>prompt diff</dt><dd class="evolution-diff">${diffTokens(A.prompt.effective?.main, B.prompt.effective?.main)}</dd>
      </dl><table class="tracks">${rows}</table></div></div>`;
  }

  // --- panels ------------------------------------------------------------
  function selected() {
    const m = members.get(state.selected);
    if (!m) { selectedPane.innerHTML = `<p class="empty">select an image</p>`; return; }
    const eff = m.prompt.effective, org = m.prompt.original;
    const doors = [];
    if (m.media.page) doors.push(`<a href="${esc(m.media.page)}">open image</a>`);
    if (eff) doors.push(`<a href="/search?q=${encodeURIComponent(eff.main)}">images like this prompt</a>`);
    if (eff && eff.prompt_id !== null && view.semantic.provider) doors.push(`<a href="/prompts/${eff.prompt_id}/neighbours?space=${encodeURIComponent(view.semantic.provider)}">prompts like this</a>`);
    if (view.doors.gallery_day) doors.push(`<a href="${esc(view.doors.gallery_day)}">this day in the gallery</a>`);
    selectedPane.innerHTML = `${m.media.thumbnail ? `<img src="${esc(m.media.thumbnail)}" alt="${esc(m.media.name)}">` : ""}
      <h2>${esc(m.ref)} · ${esc(m.media.name)}</h2>
      <p class="evolution-provenance">${esc(m.phase_ref || "")}</p>
      <h3>effective prompt</h3><pre>${esc(eff ? eff.text : "— not frozen —")}</pre>
      ${org ? `<h3>as written</h3><pre>${esc(org.text)}</pre><h3>written → ran</h3><p class="evolution-diff">${diffTokens(org.main, eff ? eff.main : "")}</p>` : `<p class="evolution-provenance">no original prompt was recorded by the generator</p>`}
      <p class="doors">${doors.join("")}</p>`;
  }

  function inspect() {
    const m = members.get(state.selected);
    if (!m) { inspector.innerHTML = ""; return; }
    const t = transitionTo.get(m.ref);
    const metric = (label, v, why) => `<dt>${label}</dt><dd title="${esc(why || "")}">${pct(v)}${v === null ? ` <small>${esc(why || "")}</small>` : ""}</dd>`;
    let html = `<dl class="metrics">${metric("written → ran", m.metrics.original_effective_cosine, m.metrics.original_effective_cosine_unavailable)}${metric("prompt ↔ image", m.metrics.text_image_cosine, m.metrics.text_image_cosine_unavailable)}</dl>`;
    if (t) {
      const changes = Object.entries(t.changes).filter(([, v]) => !(Array.isArray(v) && !v.length)).map(([k, v]) => `<tr><th>${k}</th><td>${Array.isArray(v) ? esc(v.join(", ")) : `${esc(v.from ?? "—")} → ${esc(v.to ?? "—")}`}</td></tr>`).join("");
      html += `<h3>from ${esc(t.from)}${t.phase_boundary ? " · phase boundary" : ""}</h3><dl class="metrics">${metric("prompt similarity", t.prompt_cosine, t.prompt_cosine_unavailable)}${metric("visual similarity", t.visual_cosine, t.visual_cosine_unavailable)}</dl><table class="tracks">${changes || "<tr><td>nothing else changed</td></tr>"}</table>`;
    } else if (!view.plan.sequenced) {
      html += `<p class="evolution-provenance">no transitions: the evidence does not establish an order</p>`;
    }
    const edges = view.lineage.filter((e) => e.parent === m.ref || e.child === m.ref);
    if (edges.length) html += `<h3>lineage</h3><ul class="lineage">${edges.map((e) => `<li>${esc(e.parent)} → ${esc(e.child)} <span class="kind">${esc(e.kind)}</span></li>`).join("")}</ul>`;
    inspector.innerHTML = html;
  }

  function draw() {
    ({ sequence, drift, families, lineage, compare })[state.tab]();
    root.querySelectorAll("[data-tab]").forEach((b) => b.classList.toggle("on", b.dataset.tab === state.tab));
    root.querySelectorAll("[data-ref]").forEach((el) => {
      el.classList.toggle("on", el.dataset.ref === state.selected);
      el.classList.toggle("pair", !!state.pair && state.pair.includes(el.dataset.ref) && el.dataset.ref !== state.selected);
    });
    root.querySelectorAll("[data-pair]").forEach((el) => el.classList.toggle("on", !!state.pair && el.dataset.pair === state.pair.join("|")));
    selected();
    inspect();
  }

  root.addEventListener("click", (event) => {
    const tab = event.target.closest("[data-tab]");
    if (tab) { state.tab = tab.dataset.tab; draw(); return; }
    const dot = event.target.closest("[data-pair]");
    if (dot) { state.pair = dot.dataset.pair.split("|"); state.selected = state.pair[1]; draw(); return; }
    const el = event.target.closest("[data-ref]");
    if (!el) return;
    if (event.shiftKey && state.selected && state.selected !== el.dataset.ref) { state.pair = [state.selected, el.dataset.ref]; state.tab = "compare"; }
    else state.selected = el.dataset.ref;
    draw();
  });
  draw();
})();
